runtime/slot-composer.js

```diff
@@ -49,9 +49,14 @@ class SlotComposer {
         reject(x);
       }
     }).then(slot => {
-      slot.associateWithParticle(particleSpec);
+      this._assignSlot(slotid, slot, particleSpec);
     });
   }
+  _assignSlot(slotid, slot, particleSpec) {
+    log(`_assignSlot("${slotid}")`);
+    slot.associateWithParticle(particleSpec);
+    this._slotIdByParticleSpec.set(particleSpec, slotid);
```

Comment:
in one of the changes I moved this line
this._slotIdByParticleSpec.set(particleSpec, slotid);
to ln 41, - basically as soon as the particle requests the slot, they are associated. so that when particle asks to release-slot, slotComposer could remove the pending request, if applicable.